### ISSUE TITLE
JavaSymbolResolver used in parallel

### ISSUE THREAD

**Issue by corradomio** [javaparser/javaparser#2668]
In this moment, it is not possible to use **JavaSymbolResolver** in parallel.
The problem is in **JavaParserFacade**, in the static method **get:**

    public static JavaParserFacade get(TypeSolver typeSolver) {
        return instances.computeIfAbsent(typeSolver, JavaParserFacade::new);
    }

**instances** is a **WeakHashMap** but this class is not thread-safe.
It can be enough to convert this method in a "synchronized" method:

    public static **synchronized** JavaParserFacade get(TypeSolver typeSolver) {
        return instances.computeIfAbsent(typeSolver, JavaParserFacade::new);
    }

**Comment by corradomio** [comment]
Yes, it is!
There are some changes, some classes must be reimplemented, but it is possible!
Currently I am using a ""patched"" version in parallel.

**Comment by MysterAitch** [comment]
Just wanted to flag up these really helpful / thoughtful comments from @migliorabile in a separate issue, to give some more context:

https://github.com/javaparser/javaparser/issues/2671#issuecomment-628973761

https://github.com/javaparser/javaparser/issues/2671#issuecomment-628979560

> 
> 
> At the moment, "JavaParserFacade.get()" IS ""the problem""because
> 
>     1. it is a GLOBAL/STATIC method
> 
>     2. it caches ALL type solvers
> 
>     3. it is used in several places inside the library
> 
>     4. it is NOT synchronized
> 
> 
> I tried to remove it, but I have not implemented this part.
> The current ""possible"" solution is
> 
>     1. to create a JavaParserFacadeFactory that contains the map typesolver -> JavaParserFacade (the current static part of JavaParserFacade  inserted in a separated class)
> 
>     2. to extend TypeSolver with a method "getFactory" that return a JavaParserFacadeFactory >
> 
> 
> The call (repeat, used in the library implementation)
> 
> ```java
> JavaParserFacade.get(typeSolver)
> ```
> 
> it will converted in
> 
> ```java
> public static JavaParserFacade get(TypeSolver typeSolver) {
>     return typesolver.getRoot().getFactory().get(typeSolver)
> }
> ```
> 
> In this way, the root typeSolver contains the ""thread-local"" cache of type solvers, and with a minimum impact on the current code. The problem is to change the current typeSolver interface.
> 
> But I am not sure.
